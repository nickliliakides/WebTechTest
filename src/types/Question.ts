export interface QuestionResponse {
  response_code: number;
  results?: Question[] | null;
}

export interface Question {
  id: number;
  category: string;
  type: string;
  difficulty: string;
  question: string;
  correct_answer: string;
  incorrect_answers: string[];
}
