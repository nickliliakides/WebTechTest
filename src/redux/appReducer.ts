import { createStore, combineReducers, applyMiddleware } from 'redux';

import createSagaMiddleware from 'redux-saga';

import rootSaga from './sagas/rootSaga';

import * as sportwebservice from './actions/questionwebservice';

import { Question } from '../types/Question';

export interface QuizState {
  questions: Question[];
  selectedQuestion: Question | {};
}

const initialQuizState: QuizState = {
  questions: [],
  selectedQuestion: {},
};

function questionReduxReducer(
  state = initialQuizState,
  action: { type: string; payload: any }
) {
  const { type, payload } = action;

  switch (type) {
    case sportwebservice.Types.GET_QUESTIONS_SUCCESS:
      return {
        ...state,
        questions: [...payload.results, ...state.questions],
      };
    case sportwebservice.Types.DELETE_QUESTION_REQUEST:
      return {
        ...state,
        questions: [
          ...state.questions.filter((question) => question.id !== payload),
        ],
      };
    case sportwebservice.Types.SET_SELECTED_QUESTION:
      return {
        ...state,
        selectedQuestion: payload,
      };
    case sportwebservice.Types.UPDATE_SELECTED_QUESTION:
      return {
        questions: [
          payload,
          ...state.questions.filter((question) => question.id !== payload.id),
        ],
        selectedQuestion: {},
      };
    default:
      return state;
  }
}

const overallReducer = combineReducers({
  quiz: questionReduxReducer,
});

// Sage Middleware is used for fetching external resources.
const sagaMiddleware = createSagaMiddleware();

const store = createStore(overallReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootSaga);

export { store };
