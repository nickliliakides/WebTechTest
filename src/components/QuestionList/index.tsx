import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Box, Button } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  deleteQuestionRequest,
  getQuestionsRequest,
  setSelectedQuestion,
} from '../../redux/actions/questionwebservice';
import Fade from 'react-reveal/Fade';
import EditIcon from '@material-ui/icons/Edit';
import DeleteIcon from '@material-ui/icons/Delete';
import CircularProgress from '@material-ui/core/CircularProgress';
import { Question } from '../../types/Question';
import { Dispatch } from 'redux';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      '& .MuiTextField-root': {
        margin: theme.spacing(1),
        width: '25ch',
      },
    },
    top_container: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    paper: {
      padding: theme.spacing(2),
      textAlign: 'center',
      color: theme.palette.text.secondary,
      display: 'flex',
      opacity: '.85',
      justifyContent: 'space-evenly',
    },
    questionDetails: {
      padding: theme.spacing(2),
      textAlign: 'left',
      color: theme.palette.text.secondary,
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'flex-start',
      justifyContent: 'space-between',
      flexGrow: 1,
    },
    buttons: {
      padding: theme.spacing(2),
      textAlign: 'center',
      color: theme.palette.text.secondary,
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      alignItems: 'flex-end',
    },
    button: {
      width: 100,
    },
    edit_btn: {
      backgroundColor: 'green',
      '&:hover': {
        backgroundColor: 'green',
        opacity: 0.8,
      },
    },
    delete_btn: {
      backgroundColor: 'red',
      marginTop: 10,
      '&:hover': {
        backgroundColor: 'red',
        opacity: 0.8,
      },
    },
    loader: {
      position: 'absolute',
      top: '50%',
      left: '75%',
      transform: 'translate(-50%, -50%)',
    },
  })
);

interface IProps {
  dispatch: Dispatch;
  questions?: Question[] | null;
}

const QuestionListView: React.FC<IProps> = ({
  dispatch,
  questions,
}): JSX.Element => {
  const classes = useStyles();

  useEffect(() => {
    dispatch(getQuestionsRequest());
  }, [dispatch]);

  const getMoreQuestions = (): void => {
    dispatch(getQuestionsRequest());
  };

  const deleteQuestion = (index: number): void => {
    dispatch(deleteQuestionRequest(index));
  };

  const editQuestion = (question: Question): void => {
    dispatch(setSelectedQuestion(question));
    window.scroll({ top: 0, left: 0, behavior: 'smooth' });
  };

  return questions && questions.length === 0 ? (
    <CircularProgress className={classes.loader} thickness={6} size={80} />
  ) : (
    <Grid item={true} md={5}>
      <div className={classes.top_container}>
        <h1>Questions</h1>
        <Button
          style={{ marginLeft: 12 }}
          variant='contained'
          color='primary'
          onClick={() => getMoreQuestions()}
        >
          Fetch More Questions
        </Button>
      </div>
      <Grid container spacing={3}>
        {questions &&
          questions.map((question: Question, index: number) => (
            <Grid item xs={12} key={index}>
              <Fade right duration={600} delay={70 * index}>
                <Paper className={classes.paper}>
                  <Box className={classes.questionDetails}>
                    <Box>{question.question}</Box>
                    <Box style={{ display: 'inline', fontWeight: 700 }}>
                      Category: {question.category}
                    </Box>
                    <Box style={{ fontWeight: 700 }} className='difficulty'>
                      Difficulty: {question.difficulty}
                    </Box>
                  </Box>
                  <Box className={classes.buttons}>
                    <Button
                      data-testid='edit-button'
                      className={`${classes.edit_btn} ${classes.button}`}
                      variant='contained'
                      color='primary'
                      onClick={() => editQuestion(question)}
                      startIcon={<EditIcon />}
                    >
                      Edit
                    </Button>
                    <Button
                      className={`${classes.delete_btn} ${classes.button}`}
                      variant='contained'
                      color='primary'
                      onClick={() => deleteQuestion(question.id)}
                      startIcon={<DeleteIcon />}
                    >
                      Delete{' '}
                    </Button>
                  </Box>
                </Paper>
              </Fade>
            </Grid>
          ))}
      </Grid>
    </Grid>
  );
};

const mapStateToProps = (state: { quiz: { questions: Question[] } }) => {
  return { questions: state.quiz.questions };
};

export default connect(mapStateToProps)(QuestionListView);
