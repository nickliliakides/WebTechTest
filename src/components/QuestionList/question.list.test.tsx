import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { Store, Dispatch } from 'redux';
import configureStore from 'redux-mock-store';
import QuestionList from './index';
import renderer from 'react-test-renderer';
import { Question } from '../../types/Question';

const mockStore = configureStore([]);

interface IProps {
  dispatch: Dispatch;
  questions: Question[];
}

describe('Question list tests', () => {
  let store: Store;
  let mockDispatch: jest.Mock;
  let mockProps: IProps;
  let element: JSX.Element;

  const mockQuestions = [
    {
      id: 1,
      category: 'test',
      type: 'test',
      difficulty: 'test',
      question: 'test',
      correct_answer: 'test',
      incorrect_answers: ['test', 'test', 'test'],
    },
    {
      id: 2,
      category: 'test',
      type: 'test',
      difficulty: 'test',
      question: 'test',
      correct_answer: 'test',
      incorrect_answers: ['test', 'test', 'test'],
    },
  ];

  beforeEach(() => {
    store = mockStore({
      quiz: {
        questions: mockQuestions,
      },
    });

    mockDispatch = jest.fn();

    mockProps = {
      dispatch: mockDispatch,
      questions: mockQuestions,
    };

    element = (
      <Provider store={store}>
        <QuestionList {...mockProps} />
      </Provider>
    );
  });

  it('Snapshot Testing', () => {
    const tree = renderer.create(element).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Expect to render Question list component', () => {
    const div = document.createElement('div');
    ReactDOM.render(element, div);
  });
});
