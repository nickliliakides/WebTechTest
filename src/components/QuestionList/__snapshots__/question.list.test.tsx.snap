// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question list tests Snapshot Testing 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-5"
>
  <div
    className="makeStyles-top_container-2"
  >
    <h1>
      Questions
    </h1>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginLeft": 12,
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Fetch More Questions
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="react-reveal"
        style={
          Object {
            "opacity": undefined,
          }
        }
      >
        <div
          className="MuiPaper-root makeStyles-paper-3 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            className="MuiBox-root MuiBox-root-10 makeStyles-questionDetails-4"
          >
            <div
              className="MuiBox-root MuiBox-root-11"
            >
              test
            </div>
            <div
              className="MuiBox-root MuiBox-root-12"
              style={
                Object {
                  "display": "inline",
                  "fontWeight": 700,
                }
              }
            >
              Category: 
              test
            </div>
            <div
              className="MuiBox-root MuiBox-root-13 difficulty"
              style={
                Object {
                  "fontWeight": 700,
                }
              }
            >
              Difficulty: 
              test
            </div>
          </div>
          <div
            className="MuiBox-root MuiBox-root-14 makeStyles-buttons-5"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-edit_btn-7 makeStyles-button-6 MuiButton-containedPrimary"
              data-testid="edit-button"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                Edit
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-delete_btn-8 makeStyles-button-6 MuiButton-containedPrimary"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                Delete
                 
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="react-reveal"
        style={
          Object {
            "opacity": undefined,
          }
        }
      >
        <div
          className="MuiPaper-root makeStyles-paper-3 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            className="MuiBox-root MuiBox-root-15 makeStyles-questionDetails-4"
          >
            <div
              className="MuiBox-root MuiBox-root-16"
            >
              test
            </div>
            <div
              className="MuiBox-root MuiBox-root-17"
              style={
                Object {
                  "display": "inline",
                  "fontWeight": 700,
                }
              }
            >
              Category: 
              test
            </div>
            <div
              className="MuiBox-root MuiBox-root-18 difficulty"
              style={
                Object {
                  "fontWeight": 700,
                }
              }
            >
              Difficulty: 
              test
            </div>
          </div>
          <div
            className="MuiBox-root MuiBox-root-19 makeStyles-buttons-5"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-edit_btn-7 makeStyles-button-6 MuiButton-containedPrimary"
              data-testid="edit-button"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                Edit
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-delete_btn-8 makeStyles-button-6 MuiButton-containedPrimary"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                Delete
                 
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
